### PR TITLE
Stream::cycle implementation

### PR DESCRIPTION
**Pull request by vertexclique**
* Adds `stream::cycle` implementation
* Contains an example for stream cycle.

### CODE REVIEW

File: src/stream/cycle.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+///
+/// assert_eq!(s.next().await, Some(1));
+/// assert_eq!(s.next().await, Some(2));
+/// assert_eq!(s.next().await, Some(3));
+/// assert_eq!(s.next().await, Some(1));
+/// assert_eq!(s.next().await, Some(2));
+/// #
+/// # }) }
+/// ```
+pub fn cycle<T>(items: Vec<T>) -> Cycle<T>
```
Comment (kpp) [inline src/stream/cycle.rs]:
```
pub fn cycle<I>(i: I) -> impl Stream<Item = <I as IntoIterator>::Item> where
    I: IntoIterator, 
{
    futures::stream::iter(i.into_iter().cycle())
}
```

Comment (kpp) [reply src/stream/cycle.rs]: No need to over engineer the issue. BTW I suggest you adding this method into futures-rs.

File: examples/stream-cycle.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+            total += v;
+            if total == 42 {
+                println!("Found {} the meaning of life!", total);
+                break;
+            }
+        }
+
+        Ok(())
+    })
+}
```
Comment (yoshuawuyts) [inline examples/stream-cycle.rs]: This example is cool, but I think it's probably fine to just have the doc test for now. In #131 and #129 there are quite a few more methods, and I think it'd be wise to try and keep our line count down where possible.

File: src/stream/cycle.rs (hunk 1 above)
Comment (yoshuawuyts) [reply src/stream/cycle.rs]: We want the `Cycle` counterpart to stdlib. Also we've clarified in a different issue that we do actually want our own concrete impls.

File: src/stream/cycle.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+///
+/// ```
+/// # #![feature(async_await)]
+/// # fn main() { async_std::task::block_on(async {
+/// #
+/// use async_std::prelude::*;
+/// use async_std::stream;
+///
+/// let mut s = stream::cycle(vec![1, 2, 3]);
+///
```
Comment (yoshuawuyts) [inline src/stream/cycle.rs]: In std the API is:

```rust
let a = [1, 2, 3];

let mut it = a.iter().cycle();

assert_eq!(it.next(), Some(&1));
assert_eq!(it.next(), Some(&2));
assert_eq!(it.next(), Some(&3));
assert_eq!(it.next(), Some(&1));
assert_eq!(it.next(), Some(&2));
assert_eq!(it.next(), Some(&3));
assert_eq!(it.next(), Some(&1));
```

I'd expect this API to be (with #125 landed):
```rust
let a = [1, 2, 3];

let mut it = a.into_stream().cycle();

assert_eq!(it.next().await, Some(&1));
assert_eq!(it.next().await, Some(&2));
assert_eq!(it.next().await, Some(&3));
assert_eq!(it.next().await, Some(&1));
assert_eq!(it.next().await, Some(&2));
assert_eq!(it.next().await, Some(&3));
assert_eq!(it.next().await, Some(&1));
```

Similar, but subtly different in that it's more generic, which should make it more widely applicable!

**Comment by yoshuawuyts** [comment]
#125 has landed now which means this can be modeled after std's API. @vertexclique do you have any interest in continuing with this patch in that direction?

**Comment by yoshuawuyts** [comment]
It's been a while, and I'm going to assume this PR will not be continued. Thanks for the work! Going to go ahead and close this.

**Pull request closed without merging**